src/mediator.ts

```diff
@@ -1,19 +1,28 @@
 import { BaseMediator } from "./baseMediator";
+import { IMediatorMiddleware } from "./interfaces";
 
 export class Mediator extends BaseMediator {
-    public Send(command: string, payload: any) {
+    private middlewares: IMediatorMiddleware[] = [];
+
+    public Send(command: string, payload: any): any {
         return this.Process(command, payload);
     }
 
-    public Request(query: string, payload: any) {
+    public Request(query: string, payload: any): any {
         return this.Process(query, payload);
     }
 
-    private Process(msg: string, payload: any) {
+    public RegisterMiddleware(middleware: IMediatorMiddleware): void {
+        this.middlewares.push(middleware);
+    }
+
+    private Process(msg: string, payload: any): any {
         let handler: any = super.Resolve(msg);
 
         try {
+            this.middlewares.forEach(m => m.PreProcess());
             handler.Validate(payload);
+            this.middlewares.reverse().forEach(m => m.PostProcess());
```

Comment:
this should be done after the actual invocation of the handler, i.e.

```
let result = handler.Handle(payload);
this.middlewares.reverse().forEach(m => m.PostProcess(payload, result));
return result;
```

Reply:
Woops, my bad. Btw, why isn't the actual "Handle" inside the try-block?

Reply:
hmm actually now that I looked at the code, the try-catch is quite unnecessary since it just re-throws the exception... if you want to, you can remove the try block completely 😄